test/common/testlib.py

```diff
@@ -231,23 +231,30 @@ def focus(self, selector):
     def blur(self, selector):
         self.call_js_func('ph_blur', selector)
 
-    def key_press(self, keys):
-        for k in keys:
-            if k == "Backspace":
-                self.cdp.invoke("Input.dispatchKeyEvent", type="keyDown", windowsVirtualKeyCode=8)
-            elif k.isalnum():
-                self.cdp.invoke("Input.dispatchKeyEvent", type="char", text=k, key=k)
+    def key_press(self, keys, modifiers=0):
+        for key in keys:
+            args = { "type":"keyDown", "modifiers":modifiers };
+
+            # If modifiers are used we need to pass windowsVirtualKeyCode which is
+            # basically the asci decimal representation of the key
+            args["text"] = key
+            if (not key.isalnum()  and ord(key.upper()) < 32) or modifiers != 0:
+                args["windowsVirtualKeyCode"] = ord(key.upper())
             else:
-                self.cdp.invoke("Input.dispatchKeyEvent", type="char", text=k)
+                args["key"] = key
 
-    def set_input_text(self, selector, val):
-        self.set_val(selector, "")
+            self.cdp.invoke("Input.dispatchKeyEvent", **args)
+
+    def set_input_text(self, selector, val, append=False):
         self.focus(selector)
+        if not append:
+            self.key_press("a", 2); # Ctrl + a
+
         if val == "":
-            # We need some real action for React to emit change signals
-            self.key_press([ " ", "Backspace" ])
+            self.key_press("\b"); # Backspace
```

Comment:
That looks a little odd now. I think Ctrl+A and Backspace should always go together. How about doing these `if not append or val == ""`?

Reply:
Control+A and Backspace should not always go together.
Imagine input fields where you cannot set them to empty string with Control-A + Backspace.
We have such scenario in machines, where there is input field, (for vcpu count) where the minimum number you can set is '1'. If you hit Control-A + Backspace, the input will auto fill 1, since this is the minimum. The only way to fill this input from the tests, is to use, 'append=false' mode and not use backspace, and then overwrite the input by typing over the selected text.

Reply:
ack, that makes sense, thanks!

(... and semicolon again).